Ensure AlertStore directory exists on first run

rickpasetto opened the pull request.

rickpasetto committed:
Ensure AlertStore directory exists on first run

rickpasetto committed:
Fix unit test

ps2 commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -176,6 +177,20 @@ extension AlertManager {
+        do {
+            try fileManager.createDirectory(at: documentsDirectory.appendingPathComponent(AlertStore.storageDirectoryPathComponent),
+                                            withIntermediateDirectories: true,
+                                            attributes: [FileAttributeKey.protectionKey: FileProtectionType.none])
```
This changes the dir from class C to class D, which means no protection, even before first unlock. We don't need access before first unlock, so I think this is unnecessary and more open than needed.

rickpasetto replied:
It's what the other Store's do @ps2 .  See `PersistenceController.swift`, `initializeStack` around line 149.  I'm guessing this is done so it can be accessed BFU.  Why do you say we need no access BFU?  And if we don't, shouldn't we also change `initializeStack`?  And if we do _that_, might that have other implications?

ps2 replied:
iOS should not (according to apple) launch Loop BFU.  Doing so was a bug, apparently. The other inits were written back before we realized that running BFU was not a possibility. So yes, they could be updated.

ps2 commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -43,7 +44,7 @@ public class AlertStore {
         if let storageDirectoryURL = storageDirectoryURL {
             let storageFileURL = storageDirectoryURL
-                .appendingPathComponent("AlertStore")
+                .appendingPathComponent(AlertStore.storageDirectoryPathComponent)
```
[nit] Making the location of the file backing this store solely a concern of the caller might be a little cleaner.

rickpasetto replied:
I considered that.  I am trying to avoid having the store need to make the directory for placement.

ps2 replied:
Yeah, the store would not need to worry about that if we make the storage location a concern of the caller. See 
https://github.com/tidepool-org/Loop/pull/229/commits/1d9791a263bc706391565ea4a223322e64d5b769

rickpasetto replied:
Oh, I see.  If I understand you correctly, you are not a fan of imbuing `AlertStore` with the knowledge of what subdirectory to live in.  That's fair.  I'll change that.

ps2 reviewed (changes requested):
I think the protection level isn't right here, and a small suggestion for responsibility organization.

rickpasetto committed:
Let AlertManager decide where to put AlertStore

darinkrauss commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -54,7 +54,9 @@ public final class AlertManager {
         self.fileManager = fileManager
         let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first
-        self.alertStore = alertStore ?? AlertStore(storageDirectoryURL: documentsDirectory, expireAfter: expireAfter)
+        let alertStoreDirectory = documentsDirectory?.appendingPathComponent("AlertStore")
```
What happens to the overall functionality if `documentsDirectory` is nil? Would it better to just force unwrap it on the line above? That is, `.first!`?

rickpasetto replied:
No.  This is intentional, because for Unit testing it could be nil

darinkrauss commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -209,6 +211,23 @@ extension AlertManager {
+            return
+        }
+        do {
+            var isDirectory: ObjCBool = false
```
Why not just replace this entire function with this in the AlertManager?
```
do {
  try fileManager.createDirectory(at: alertStoreDirectory, withIntermediateDirectories: true, attributes: nil)
} catch let error {
  log.error("Could not create alert store directory: %{public}@", error)
}
```

rickpasetto replied:
Because this one is a little more thorough?  Not sure I understand the question.

darinkrauss replied:
What is more thorough? I must be missing something.

darinkrauss reviewed (commented):
A couple of questions/comments.

ps2 commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -224,6 +211,23 @@ extension AlertManager {
 extension FileManager {
+    
+    fileprivate func ensureDirectoryExists(_ directory: URL?, log: DiagnosticLog) {
+        guard let directory = directory else {
```
[nit] Make `directory` non-optional? Seems odd to be calling a method named `ensureDirectoryExists` and pass it nil for the directory.

rickpasetto replied:
It's for unit testing

ps2 commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -224,6 +211,23 @@ extension AlertManager {
 
 extension FileManager {
+    
+    fileprivate func ensureDirectoryExists(_ directory: URL?, log: DiagnosticLog) {
```
[nit] maybe return success/failure, so caller can decide what to do in those cases?

ps2 commented on Loop/Managers/Alerts/AlertManager.swift:
```diff
@@ -224,6 +211,23 @@ extension AlertManager {
+        do {
+            var isDirectory: ObjCBool = false
+            let exists = fileExists(atPath: directory.absoluteString, isDirectory: &isDirectory)
+            if !(exists && isDirectory.boolValue) {
```
Why try calling createDirectory if the path exists and is a file?

rickpasetto replied:
I figured it was more thorough.  In practice this shouldn't be a problem, but if it ever happened I wanted the failure to be clearer (because it wouldn't create a directory on top of a file)

ps2 replied:
```

ps2 reviewed (approved):
LGTM. Just some nits.

rickpasetto committed:
Moved functions to LoopKit

I've consolidated some of the FileManager functions into LoopKit, where they can now be shared (and made more consistent)

rickpasetto commented:
Based on your feedback, @ps2 and @darinkrauss, I've moved some of these `FileManager` functions to LoopKit (https://github.com/tidepool-org/LoopKit/pull/236).  Please LMK if this addresses your concerns.

rickpasetto committed:
PR Feedback: don't use public FileManager extension

rickpasetto commented:
@ps2 and @darinkrauss ok, based on @ps2 's feedback to me in Slack, I have "de-publicized" the `FileManager` extension in LoopKit (see https://github.com/tidepool-org/LoopKit/pull/236 -- I'd forgotten about the rule that we shouldn't expose public platform extensions in LoopKit), and copied the code here.  Please have a look, thanks!

ps2 reviewed (approved):
LGTM

rickpasetto committed:
Merge branch 'dev' into rick/ensure-alert-store-directory-exists
